File: SIH/cyber-risk-platform/backend/app/schemas/asset.py
import uuid
from pydantic import BaseModel, ConfigDict, Field, IPvAnyAddress, field_validator
from typing import Any
from datetime import datetime
import re

class AssetBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    asset_type: str = Field(..., max_length=100)
    environment: str = Field(..., max_length=100)
    criticality: int = Field(default=0, ge=0, le=100)
    business_value: float = Field(default=0.0, ge=0.0)
    owner: str | None = Field(default=None, max_length=255)
    department: str | None = Field(default=None, max_length=255)
    hostname: str | None = Field(default=None, max_length=255)
    ip_address: IPvAnyAddress | str | None = None
    operating_system: str | None = Field(default=None, max_length=255)
    technology: str | None = Field(default=None, max_length=255)
    internet_exposed: bool = False
    status: str = Field(default="active", max_length=50)

    @field_validator("hostname")
    @classmethod
    def validate_hostname(cls, v: str | None) -> str | None:
        if v is not None:
            if not re.match(r"^[a-zA-Z0-9.-]+$", v):
                raise ValueError("Invalid hostname format")
        return v

    @field_validator("ip_address")
    @classmethod
    def validate_ip(cls, v: Any) -> str | None:
        if v is not None:
            return str(v)
        return v


class AssetCreate(AssetBase):
    organization_id: uuid.UUID

class AssetUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    asset_type: str | None = Field(default=None, max_length=100)
    environment: str | None = Field(default=None, max_length=100)
    criticality: int | None = Field(default=None, ge=0, le=100)
    business_value: float | None = Field(default=None, ge=0.0)
    owner: str | None = Field(default=None, max_length=255)
    department: str | None = Field(default=None, max_length=255)
    hostname: str | None = Field(default=None, max_length=255)
    ip_address: IPvAnyAddress | str | None = None
    operating_system: str | None = Field(default=None, max_length=255)
    technology: str | None = Field(default=None, max_length=255)
    internet_exposed: bool | None = None
    status: str | None = Field(default=None, max_length=50)

    @field_validator("hostname")
    @classmethod
    def validate_hostname(cls, v: str | None) -> str | None:
        if v is not None:
            if not re.match(r"^[a-zA-Z0-9.-]+$", v):
                raise ValueError("Invalid hostname format")
        return v

    @field_validator("ip_address")
    @classmethod
    def validate_ip(cls, v: Any) -> str | None:
        if v is not None:
            return str(v)
        return v


class AssetRead(AssetBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class PaginatedAssetsResponse(BaseModel):
    items: list[AssetRead]
    page: int
    page_size: int
    total: int
    total_pages: int


class AssetPostureResponse(BaseModel):
    asset_id: uuid.UUID
    asset_name: str
    criticality: int
    internet_exposed: bool
    open_vulnerabilities: int
    critical_vulnerabilities: int
    recent_telemetry_events: int

File: SIH/cyber-risk-platform/frontend/src/types/asset.ts
export type AssetType = 'server' | 'database' | 'application' | 'endpoint' | 'network_device' | 'cloud_resource' | 'api' | 'payment_system' | 'storage' | 'container' | 'virtual_machine' | 'other';
export type AssetEnvironment = 'production' | 'staging' | 'development' | 'testing';
export type AssetStatus = 'active' | 'inactive' | 'maintenance' | 'retired';

export interface Asset {
  id: string;
  organization_id: string;
  name: string;
  description?: string;
  asset_type: AssetType | string;
  environment: AssetEnvironment | string;
  criticality: number; // 0-100
  business_value: number;
  owner?: string;
  department?: string;
  hostname?: string;
  ip_address?: string;
  operating_system?: string;
  technology?: string;
  internet_exposed: boolean;
  status: AssetStatus | string;
  created_at: string;
  updated_at: string;
}

export interface AssetCreateRequest {
  organization_id: string;
  name: string;
  description?: string;
  asset_type: string;
  environment: string;
  criticality: number;
  business_value: number;
  owner?: string;
  department?: string;
  hostname?: string;
  ip_address?: string;
  operating_system?: string;
  technology?: string;
  internet_exposed: boolean;
  status: string;
}

export type AssetUpdateRequest = Partial<Omit<AssetCreateRequest, 'organization_id'>>;

export interface PaginatedAssets {
  items: Asset[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface AssetFilters {
  search?: string;
  organization_id?: string;
  asset_type?: string;
  environment?: string;
  status?: string;
  internet_exposed?: boolean;
  criticality_min?: number;
  criticality_max?: number;
  department?: string;
  owner?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
}

export interface AssetPosture {
  asset_id: string;
  asset_name: string;
  criticality: number;
  internet_exposed: boolean;
  open_vulnerabilities: number;
  critical_vulnerabilities: number;
  recent_telemetry_events: number;
}
